src/proxy/connection_manager.rs

```diff
@@ -63,16 +68,110 @@ impl Default for ConnectionManager {
     fn default() -> Self {
         ConnectionManager {
             drains: Arc::new(RwLock::new(HashMap::new())),
+            outbound_connections: Arc::new(RwLock::new(HashSet::new())),
+        }
+    }
+}
+
+pub struct ConnectionGuard {
+    cm: ConnectionManager,
+    ctx: ProxyRbacContext,
+    watch: Option<drain::Watch>,
+}
+
+impl ConnectionGuard {
+    pub async fn handle_connection(
+        mut self,
+        send: impl Future<Output = Result<(u64, u64), Error>> + Sized,
+    ) -> Result<(u64, u64), Error> {
+        let watch = self.watch.take().expect("watch cannot be taken twice");
```

Comment:
comment would be nice, it's a little subtle why this expect is OK I think.
"watch can't be taken twice because this ConnectionGuard is consumed by handle_connection" maybe

Reply:
Shouldn't we error instead of crashing here?

Reply:
I think because the guard is owned by handle_connection and not returned it shouldn't ever be called twice. It's dropped once this returns.